src/util/http/websocket/QueryHub.cpp

```diff
@@ -8,71 +8,35 @@
   }
 
-  auto pointerHolder =
-      std::make_shared<std::optional<PointerType>>(std::nullopt);
-  auto cleanupAction = [this, queryId, pointerHolder]() {
-    AD_CORRECTNESS_CHECK(pointerHolder->has_value());
-    auto referencePointer = std::move(pointerHolder->value());
-    // Only erase object if we created it, otherwise we have a race condition
-    if (socketDistributors_.contains(queryId) &&
-        equals(referencePointer, socketDistributors_.at(queryId))) {
-      socketDistributors_.erase(queryId);
-    }
-  };
-  auto distributor = std::make_shared<QueryToSocketDistributor>(
-      ioContext_, [this, cleanupAction = std::move(cleanupAction)]() mutable {
-        auto future =
-            net::dispatch(globalStrand_,
-                          std::packaged_task<void()>(std::move(cleanupAction)));
-        // If we blocked until the cleanup action is executed on different
-        // strand, this could theoretically lead to deadlocks when globalStrand_
-        // is on a different execution context. This is currently not the case,
-        // but it might be in the future, resulting in deadlocks at distance
-        // which is bad. Unfortunately destructors can't be coroutines, which
-        // would just allow us to co_await execution without issues. To prevent
-        // deadlocks, we wait at most 1 second (arbitrarily chosen interval)
-        // for the cleanup to complete, then print a warning, and wait for 9
-        // more seconds before we call std::terminate instead of running into
-        // a deadlock.
-        auto status = future.wait_for(std::chrono::seconds(1));
-        if (status != std::future_status::ready) {
-          LOG(WARN)
-              << "Distributor cleanup in the websocket module taking longer "
-                 "than usual. This might be due to heavy load or a deadlock."
-              << std::endl;
-        }
-        status = future.wait_for(std::chrono::seconds(9));
-        if (status != std::future_status::ready) {
-          LOG(FATAL)
-              << "Distributor cleanup in the websocket module taking longer "
-                 "than 10 seconds. Terminating to avoid potential deadlock."
-              << std::endl;
-          std::terminate();
-        }
+  auto distributor =
+      std::make_shared<QueryToSocketDistributor>(ioContext_, [this, queryId]() {
+        auto future = net::dispatch(net::bind_executor(
+            singleThreadPool_, std::packaged_task<void()>([this, &queryId]() {
+              AD_CORRECTNESS_CHECK(socketDistributors_.erase(queryId));
+            })));
+        // Usually this is an anti-pattern because this blocks the `io_context`
+        // preventing it from doing any other work in the meantime. However,
+        // because `singleThreadPool_` has a dedicated thread just to access
+        // `socketDistributors_` this time will usually be very short and is
+        // guaranteed to never cause a deadlock.
+        future.wait();
```

Comment:
I am still thinking about how hard a `contribute to progress on strand` function would be, but definitely not for this PR.

Reply:
But you can avoid the temporary `future` variable, by calling `wait()` directly on the temporary.

Reply:
> I am still thinking about how hard a contribute to progress on strand function would be, but definitely not for this PR.

I don't think there's a general solution to this problem, but at least for io context you could use a thread local to store a pointer to the io_context before calling `ioContext.run()`, then check if this thread_local is set so you can call `run_one()` in a loop until the future is ready. For other execution_contexts I don't think there's a way.

Reply:
I just read about this. But this is again only useful if your threadLocal `ioContext` is the same one that is also associated with the `strand` you are dispatching to (or maybe not, in this case it might be that you just make progress for someone else, which is also not too bad).
But probably you could also end up waiting for a long time if that `single handler` consists of an infinite loop of ready `dispatch` calls, am I right?
Nothing of this needs to be implemented here, but we are in the process of understanding things.